app/src/org/commcare/dalvik/odk/provider/InstanceProvider.java

```diff
@@ -365,10 +386,94 @@ public int update(Uri uri, ContentValues values, String where, String[] whereArg
                 throw new IllegalArgumentException("Unknown URI " + uri);
         }
 
+        // If we've changed a particular form instance's status, and not
+        // created a new entry (hence count > 0 check), we need to mirror the
+        // change in that form's record.  NOTE: this conditional is crucial,
+        // since updating a form record in turn calls this update function, so
+        // we need to break the infinite loop by only updating the form record
+        // when the the status changes.
+        if (values.containsKey(InstanceColumns.STATUS) && count > 0) {
+            try {
+                linkToSessionFormRecord(uri);
+            } catch (Exception e) {
+                // TODO: correctly exit function with error code so that
+                // callers can propogate the failure
+            }
+        }
+
         getContext().getContentResolver().notifyChange(uri, null);
+
         return count;
     }
 
+    /**
+     * Register instance with the session's form record.
+     *
+     * @param instanceUri points to a concrete instance we want to register
+     */
+    private void linkToSessionFormRecord(Uri instanceUri) {
+        AndroidSessionWrapper currentState = CommCareApplication._().getCurrentSessionWrapper();
+
+        if (instanceUri == null) {
+            raiseFormEntryError("Form Entry did not return a form", currentState);
+            return;
+        }
+
+        Cursor c = getContext().getContentResolver().query(instanceUri, null, null, null, null);
+        boolean complete = false;
+        try {
+            // register the instance uri and its status with the session
+            // NOTE: there is no safety checks preventing us from binding two
+            // different form instances to the form record found in the
+            // session. We just need to be careful to flush the session's form
```

Comment:
I don't think this should be affected.

This is basically just noting that the instance URI needs to uniquely represent a single form record, since we assume there's a one-to-one parity.